### PR TITLE
cling, xeus-cling: vendor llvm

### PR DESCRIPTION
**Pull request by wegank**
## Description of changes

This PR attempts to build Cling with root's LLVM fork, as the packages involved are the last blockers for the removal of LLVM 8-11 in Nixpkgs (see #283954). The refactor basically follows https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=cling.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by lorenz** [comment]
FYI there seems to be a small patchset which allows using LLVM 17 with xeus-cling: https://github.com/jupyter-xeus/xeus-cling/compare/main...lackhole:xeus-cling:llvm-16

**Comment by wegank** [comment]
> FYI there seems to be a small patchset which allows using LLVM 17 with xeus-cling: [jupyter-xeus/xeus-cling@main...lackhole:xeus-cling:llvm-16](https://github.com/jupyter-xeus/xeus-cling/compare/main...lackhole:xeus-cling:llvm-16)

I guess this won't work unless Cling is based on LLVM 16...

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 283989` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>cling</li>
    <li>xeus-cling</li>
  </ul>
</details>

**Comment by thomasjm** [comment]
A few thoughts here, it's been a few years so I don't remember everything perfectly:

* Cling is pretty tightly bound to a given LLVM release, in this case LLVM 9. I don't think you're going to have luck trying to build Cling with a totally different LLVM version.
* Building with Root's LLVM fork could potentially work, but why? You'll just have replaced the current Nixpkgs build of LLVM 9 with another (slightly weird/patched) version.
* Over its history, Cling has sometimes required patches to both LLVM and Clang. IIRC, an earlier version of this derivation used a vendored LLVM. But this was not preferred, because then Hydra had to build another whole other copy of LLVM. It's more efficient to use Nixpkgs' own LLVM build when possible.

In short, I don't think it'll be easy to get rid of LLVM 9. It might be worth getting in touch with the Cling team to see if a new release might be coming soon.

**Comment by a-n-n-a-l-e-e** [comment]
Result of `nixpkgs-review pr 283989` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>cling</li>
    <li>xeus-cling</li>
  </ul>
</details>

---

Result of `nixpkgs-review pr 283989` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>cling</li>
    <li>xeus-cling</li>
  </ul>
</details>

**Comment by a-n-n-a-l-e-e** [comment]
> A few thoughts here, it's been a few years so I don't remember everything perfectly:
> 
>     * Cling is pretty tightly bound to a given LLVM release, in this case LLVM 9. I don't think you're going to have luck trying to build Cling with a totally different LLVM version.

it builds -- do you have any suggestions on a good way to test? note that the current darwin build is generating errors and this PR seems to get further but generates different errors. linux can include iostream and write to cout. that's the extent i've tested.

>     * Building with Root's LLVM fork could potentially work, but why? You'll just have replaced the current Nixpkgs build of LLVM 9 with another (slightly weird/patched) version.

there is a maintenance burden. LLVM9 is different enough from newer LLVMs that the toolchain requires a bunch of special cases for everything to work. there are some darwin bugs with the current toolchains i'm working on and not having to support older toolchains will make the changes and the code simpler. there is a cost to just leaving things as is. 

> 
>     * Over its history, Cling has sometimes required patches to both LLVM and Clang. IIRC, an earlier version of this derivation used a vendored LLVM. But this was not preferred, because then Hydra had to build another whole other copy of LLVM. It's more efficient to use Nixpkgs' own LLVM build when possible.

given cling is the only package using LLVM9 it already is requiring Hydra to build a whole other LLVM.

> In short, I don't think it'll be easy to get rid of LLVM 9. It might be worth getting in touch with the Cling team to see if a new release might be coming soon.

a new version was released in december.

File: pkgs/development/interpreters/cling/default.nix [hunk 1]
```diff
@@ -34,36 +34,37 @@
...
 
-  # The LLVM 9 headers have a couple bugs we need to patch
-  fixedLlvmDev = runCommand "llvm-dev-${llvmPackages_9.llvm.version}" { buildInputs = [git]; } ''
-    mkdir $out
-    cp -r ${llvmPackages_9.llvm.dev}/include $out
-    cd $out
-    chmod -R u+w include
-    git apply ${./fix-llvm-include.patch}
-  '';
+  version = "0.9";
```
Comment (thomasjm) [inline pkgs/development/interpreters/cling/default.nix:37]: This should probably be `1.0` to pick up the latest support. We may need to ping the Cling team to get them to cut a Git release with a tag.

**Comment by thomasjm** [comment]
> a new version was released in december.

Oh, really? They didn't tag it on GitHub. That would change things.

I went digging around and found this: https://github.com/vgvassilev/cling/blob/v1.0/docs/ReleaseNotes.md. It indicates you can build Cling with LLVM 13.

Somewhat confusingly, the build instructions on [this page](https://root.cern/cling/cling_build_instructions/) mention a LLVM branch called `cling-llvm16` :thinking: 

> do you have any suggestions on a good way to test?

A good start would be to run the full sample Jupyter notebook provided by `xeus-cling`.

**Comment by wegank** [comment]
They did [tag the release](https://github.com/root-project/cling/releases/tag/v1.0), it was just strangely oddly in vgvassilev's repo.

Also, the official instruction is always to [build in-tree with their own fork of LLVM](https://github.com/root-project/cling/tree/v1.0#building-from-source), and building with `llvmPackages_13` is likely broken and unsupported anyway (vgvassilev/cling#230).

**Comment by a-n-n-a-l-e-e** [comment]
> > a new version was released in december.
> 
> Oh, really? They didn't tag it on GitHub. That would change things.
> 
> I went digging around and found this: https://github.com/vgvassilev/cling/blob/v1.0/docs/ReleaseNotes.md. It indicates you can build Cling with LLVM 13.

I was messing around yesterday and put this together yesterday. i got it to build and hello world worked but was pulling header files from my system so still needed work.
<details>
<summary>default.nix</summary>
<pre>
{ lib
, stdenv
, cmake
, fetchFromGitHub
, ninja
, python3
, ncurses
, zlib
}:

let
  clingRepo = "cling";
  llvmRepo = "llvm-project";
in
stdenv.mkDerivation (finalAttrs: {
  pname = "cling2"; 
  version = "1.0";
  
  srcs = let owner = "root-project"; in [
    (fetchFromGitHub {
      inherit owner; 
      repo = clingRepo;
      rev = "refs/tags/v${finalAttrs.version}";
      hash = "sha256-Ye8EINzt+dyNvUIRydACXzb/xEPLm0YSkz08Xxw3xp4=";
      name = clingRepo;
    })
    (fetchFromGitHub {
      inherit owner; 
      repo = llvmRepo;
      #rev = "refs/tags/cling-llvm16-20231130-01";
      #hash = "sha256-h/cstR1z6SLooVcNUvT2FsY53bBsZbU81Wwi9AUZbOIr";
      rev = "refs/tags/cling-llvm13-20231016-01";
      hash = "sha256-Unw5RzYcVYJsViaM9t+QhuBXzTY/r6qo9TDzwZkFi5c=";
      name = llvmRepo;
    })
  ];
  
  sourceRoot = ".";
  
  postUnpack = ''
    chmod u+w -R .
  '';
  
  strictDeps = true;
  
  nativeBuildInputs = [
    cmake
    ninja
    python3
  ];
  
  buildInputs = [
    ncurses
    zlib
  ];
  
  cmakeDir = "../${llvmRepo}/llvm";
  
  cmakeFlags = [
    "-DLLVM_BUILD_TOOLS=OFF"
    "-DLLVM_ENABLE_PROJECTS=clang"
    "-DLLVM_EXTERNAL_CLING_SOURCE_DIR=../${clingRepo}"
    "-DLLVM_EXTERNAL_PROJECTS=cling"
    "-DLLVM_TARGETS_TO_BUILD=host;NVPTX"
  ];

  meta = with lib; {
    description = "The Interactive C++ Interpreter";
    homepage = "https://root.cern/cling/";
    license = with licenses; [ lgpl21 ncsa ];
    maintainers = with maintainers; [ thomasjm ];
    platforms = platforms.unix;
  };
})  
</pre>
</details>

> Somewhat confusingly, the build instructions on [this page](https://root.cern/cling/cling_build_instructions/) mention a LLVM branch called `cling-llvm16` 🤔

using the latest cling rev on main and the the latest cling-llvm16 from the llvm fork will build in the above derivation and hello world works.

I saw @wegank working on this PR and was hoping it would solve all our problems.

> > do you have any suggestions on a good way to test?
> 
> A good start would be to run the full sample Jupyter notebook provided by `xeus-cling`: https://github.com/jupyter-xeus/xeus-cling/blob/main/notebooks/xcpp.ipynb.

thanks!

**Comment by wegank** [comment]
> using the latest cling rev on main and the the latest cling-llvm16 from the llvm fork will build in the above derivation and hello world works.

Oh, great! Would you like to create a PR? I'll close mine in favour of yours.

**Comment by thomasjm** [comment]
Usually when there's uncertainty about how to build I've pinged some Cling people. Historically this has been @vgvassilev or @Axel-Naumann -- any chance one of you could help us pick which version of LLVM to use with Cling these days?

Like I mentioned, if we have to choose between a vendored LLVM and using one that Nixpkgs contains anyway, we usually prefer the latter.

**Comment by thomasjm** [comment]
> the official instruction is always to [build in-tree with their own fork of LLVM](https://github.com/root-project/cling/tree/v1.0#building-from-source), and building with llvmPackages_13 is likely broken and unsupported anyway (https://github.com/vgvassilev/cling/issues/230).

Ah I see. Well, we've gotten by with unpatched LLVM 9 so far `¯\_(ツ)_/¯`. IIRC we did that with Cling maintainer blessing but maybe things have changed.

**Comment by a-n-n-a-l-e-e** [comment]
> > using the latest cling rev on main and the the latest cling-llvm16 from the llvm fork will build in the above derivation and hello world works.
> 
> Oh, great! Would you like to create a PR? I'll close mine in favour of yours.

it's not very far along. I can make it a WIP called `cling-ng` but your PR seems like it is further along. i haven't tried to integrate with xeus-cling.

**Comment by a-n-n-a-l-e-e** [comment]
https://github.com/NixOS/nixpkgs/pull/284167

**Comment by vgvassilev** [comment]
We should use cling v1.0 and llvm13. In principle we can use vanilla llvm but we need our clang because it contains patches that make cling work.

**Comment by thomasjm** [comment]
Thanks @vgvassilev!

Could you tell me where to find the necessary version of Clang? In the past I've used the tagged commits from the repo at http://root.cern/git/clang.git. But this lacks a `cling-v1.0` tag -- the latest available is `cling-v0.9`.

Also, would it be possible to create a GitHub release for the new Cling 1.0 on https://github.com/root-project/cling ?

**Comment by vgvassilev** [comment]
> Thanks @vgvassilev!
> 
> Could you tell me where to find the necessary version of Clang? In the past I've used the tagged commits from the repo at http://root.cern/git/clang.git. But this lacks a `cling-v1.0` tag -- the latest available is `cling-v0.9`.

We have switched to the llvm monorepo: https://github.com/root-project/llvm-project/ the tag that goes with v1.0 is cling-llvm13. @hahnjo, we might need to create an alias for this tag `cling-v1.0` which will have more meaningful relationship between the cling version and the llvm version...

> 
> Also, would it be possible to create a GitHub release for the new Cling 1.0 on https://github.com/root-project/cling ?

Done.

**Comment by RossComputerGuy** [comment]
I'm wondering what the progress on this is, I see the linux targets pass but not the darwin ones.

**Comment by wegank** [comment]
The darwin ones should build fine, they were failing on ofborg due to timeout.

**Comment by thomasjm** [comment]
The Linux one is not working yet, it builds but still has some include path issues to resolve.

**Comment by thomasjm** [comment]
Okay, turned out to be easier than I thought to fix the flags up. https://github.com/NixOS/nixpkgs/pull/284865 is ready for review now.

**Comment by wegank** [comment]
Superseded by #284865.
